src/Database/PG/Query/Pool.hs

```diff
@@ -158,19 +161,22 @@ withConn :: ( MonadIO m
             )
          => PGPool
          -> TxMode
+         -> e
          -> (PGConn -> ExceptT e m a)
          -> ExceptT e m a
-withConn pool txm f =
-  catchConnErr action
+withConn pool txm timeoutError f =
+  catchConnErr action timeoutError
   where
     action  = withExpiringPGconn pool $
              \connRsrc -> runTxOnConn connRsrc txm f
 
 catchConnErr :: (FromPGConnErr e, MonadError e m, MonadBaseControl IO m)
-             => m a
+             => m (Maybe a)
+             -> e
              -> m a
-catchConnErr action =
-  control $ \runInIO -> runInIO action `catch` (runInIO . handler)
+catchConnErr action timeoutError =
+  control (\runInIO -> runInIO action `catch` (runInIO . handler))
+      >>= maybe (throwError timeoutError) pure
```

Comment:
Is there a reason we don't want to `throwError timeoutError` on the left-hand side of `catch`? Let's at least document what this does. We can also remove it from the export list if it's unused and kind of arbitrary at this point